cli/2024-06-04-aws-irsa-support.md

```diff
@@ -0,0 +1,326 @@
+>default 
+
+% rad init --full  
+Select your cloud provider                     
+1. Azure                                  
+> 2. AWS                                     
+3.[back] 
+
+% rad init --full  
+
+Select the identity for the AWS cloud provider 
+1. access key                              
+> 2. IRSA  
+
+% rad init --full  
+
+Enter the AWS RoleARN for configuring IRSA: <role-arn-radius>
+
+```
+
+
+Once radius is installed, we use rad env update to update the aws provider config
+```
+% rad env update --aws-account-id <aws-account-id> --aws-region <aws-region> 
+```
+
+**Sample Output:**
+
+n/a
+
+**Sample Recipe Contract:**
+
+n/a
+
+### CLI Design
+
+`rad init --full` will need to be updated to ask the user to choose between AWS access keys and IRSA. rad credential register aws should now support two credential types, and associated parameters.
+
+We should update the Radius Helm chart to allow the user to enable IRSA during installation with the `rad install kubernets --set global.aws.IRSA.enabled=true` value. This must be passed to ucp and rp charts through values and when the value is set the pod spec should mount the service-account token.
+
+### Multi-tenancy
+
+Currently, we support only one credential. However, for future, we might want to support multiple credentials potentially across AWS accounts ( say, one for QA team and other for Dev). The code changes for IRSA should be compatible with these changes in future. Since IRSA setup establishes trust between the cluster - oidc provider and AWS, once we support multiple credentials, we should be able to assume different roleARN based on which environment we are deploying from. We just have to establish right trust entities in all roles.  
+
+### Implementation Details
+
+#### UCP
+
+UCP is responsible for communicating with AWS to deploy AWS resources. 
+UCP should be able to assume the role specified in the credential and use this to communicate with AWS since it has the service-token mounted.
+
+#### Bicep
+
+NA
+
+#### Deployment Engine
+
+NA
+
+#### Core RP  / Recipes RP
+
+Terraform provider requires AWS credentials in order to deploy recipes.
+
+RP should be able to assume the role specified in the credential and use this to communicate with AWS since its pod will have the service-token mounted.
+
+### Error Handling
+
+1. handle mismatch of credential type and arguments supplied.
+2. what if the cluster does not have a oidc provider configured?
+3. what if the user has aws credentials (secrets) configured and wants to switch to IRSA? we can print a informational message requesting user to delete the old credentials since we currently support only one credential. 
+
+## Test plan
+
+We should add a E2E to test the deployment of AWS resources using IRSA. 
+
+## Security
+
+
+
+## Compatibility
+
+
+## Monitoring and Logging
+
+We will have the same monitoring and logging as today. We will not be adding any new instrumentation.
+
+## Development plan
```

Comment:
can this work be parallelized?

Reply:
Yes. Once the model changes are finalized and merged, we should be able to parallelize the efforts (tasks number indicate sequence, same-numbered tasks can be parallelized)
1. model changes get finalized and merged
2. changes to support UCP - AWS resource deployment
2. `rad install` changes to add helm chart configurations
2. `rad credential` changes
3. changes to RP to allow terraform recipes deployment (depends on UCP to fetch credential like @kachawla pointed out)
4. `rad init` changes

we would want to merge all the cli changes in only after the server-side support is complete.